Failure to load gRPC Node module leads to deleting all resources

A user hit a case where `pulumi update --yes` deleted all resources after hitting the error below.  Presumably this somehow led the engine to not register any resources, but to believe the program had completed succesfully, so that it treated all old resources as needing to be deleted.

Clearly we should never allow this to happen.

```
Diagnostics:
  pulumi:pulumi:Stack (compute-b-compute-b-stage-eus2):
    (node:97) UnhandledPromiseRejectionWarning: Error: Failed to load gRPC binary module because it was not installed for the current system
    Expected directory: node-v64-linux-x64-glibc
    Found: [node-v64-darwin-x64-unknown]
    This problem can often be fixed by running "npm rebuild" on the current system
    Original error: Cannot find module '/data/pulumi/node_modules/grpc/src/node/extension_binary/node-v64-linux-x64-glibc/grpc_node.node'
        at Object.<anonymous> (/data/pulumi/node_modules/grpc/src/grpc_extension.js:53:17)
        at Module._compile (internal/modules/cjs/loader.js:689:30)
        at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
        at Module.load (internal/modules/cjs/loader.js:599:32)
        at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
        at Function.Module._load (internal/modules/cjs/loader.js:530:3)
        at Module.require (internal/modules/cjs/loader.js:637:17)
        at require (internal/modules/cjs/helpers.js:22:18)
        at Object.<anonymous> (/data/pulumi/node_modules/grpc/src/client_interceptors.js:144:12)
        at Module._compile (internal/modules/cjs/loader.js:689:30)
    (node:97) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)                                                                                                                 
    (node:97) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.                                                                                                                                                             

Resources:
    - 3 to delete
```

Hi, this was me who reported it.  OK.. so the host is Mac OSX 10.14.1, pulumi version on host is 0.16.8

I have a docker container built on ubuntu 18.04, also pulumi 0.16.8.

If I update my npm on the Mac host then enter the container (mounting the same filesystem from host) and try an update, this is what happens:

````
    # (enter into docker container - passing in env vars)
    
    root@c522c106993d:/data# pulumi version
    v0.16.8
    
    root@c522c106993d:/data# cd pulumi/infra/azure/compute-api/                                                                                                                                                                                                                                                                
    root@c522c106993d:/data/pulumi/infra/azure/compute-api# pulumi stack select                                                                                                                                                                                                                                                
    Logging in using access token from PULUMI_ACCESS_TOKEN                                                                                                                                                                                                                                                                        
    Please choose a stack, or create a new one: compute-api-b-stage-eus2                                                                                                                                                                                                                                     
    root@c522c106993d:/data/pulumi/infra/azure/compute-api# pulumi up --yes
    Previewing update (compute-api-b-stage-eus2):
    
         Type                                   Name                                         Plan       Info
     -   pulumi:pulumi:Stack                    compute-api-compute-api-b-stage-eus2   delete     17 messages
     -   ├─ azure:lb:Rule                       Api-B-Tcp-80-Stage-Eus2-Plb-Rule          delete
     -   ├─ azure:lb:Rule                       Api-B-Tcp-443-Stage-Eus2-Plb-Rule         delete
     -   ├─ azure:lb:BackendAddressPool         Api-B-Stage-Eus2-Plb-Pool                 delete
     -   ├─ azure:lb:Probe                      Api-B-Http-4422-Up-Stage-Eus2-P-Probe     delete
     -   ├─ azure:network:NetworkSecurityRule   Api-B-Tcp-443-Allow-Stage-Eus2-Nsg-Rule   delete
     -   ├─ azure:network:NetworkSecurityRule   Api-B-Tcp-80-Allow-Stage-Eus2-Nsg-Rule    delete
     -   ├─ azure:lb:LoadBalancer               Api-B-Stage-Eus2-Plb                      delete
     -   ├─ azure:network:NetworkSecurityRule   Api-B-Any-3444-Allow-Stage-Eus2-Nsg-Rule  delete
     -   ├─ azure:network:PublicIp              Api-B-Stage-Eus2-Pip                      delete
     -   ├─ azure:network:NetworkSecurityGroup  Api-B-Stage-Eus2-Nsg                      delete
     -   └─ azure:core:ResourceGroup            Api-B-Stage-Eus2-Rg                       delete
    
    Diagnostics:
      pulumi:pulumi:Stack (compute-api-compute-api-b-stage-eus2):
        (node:67) UnhandledPromiseRejectionWarning: Error: Failed to load gRPC binary module because it was not installed for the current system
        Expected directory: node-v64-linux-x64-glibc
        Found: [node-v64-darwin-x64-unknown]
        This problem can often be fixed by running "npm rebuild" on the current system
        Original error: Cannot find module '/data/pulumi/node_modules/grpc/src/node/extension_binary/node-v64-linux-x64-glibc/grpc_node.node'
            at Object.<anonymous> (/data/pulumi/node_modules/grpc/src/grpc_extension.js:53:17)
            at Module._compile (internal/modules/cjs/loader.js:689:30)
            at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
            at Module.load (internal/modules/cjs/loader.js:599:32)
            at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
            at Function.Module._load (internal/modules/cjs/loader.js:530:3)
            at Module.require (internal/modules/cjs/loader.js:637:17)
            at require (internal/modules/cjs/helpers.js:22:18)
            at Object.<anonymous> (/data/pulumi/node_modules/grpc/src/client_interceptors.js:144:12)
            at Module._compile (internal/modules/cjs/loader.js:689:30)
        (node:67) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)                                                                    
        (node:67) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.                                                                                                                
    
    Resources:
        - 12 to delete
    
    Updating (compute-api-b-stage-eus2):
    
         Type                                   Name                                         Status      Info
     -   pulumi:pulumi:Stack                    compute-api-compute-api-b-stage-eus2   deleted     17 messages
     -   ├─ azure:lb:Rule                       Api-B-Tcp-443-Stage-Eus2-Plb-Rule         deleted
     -   ├─ azure:lb:Rule                       Api-B-Tcp-80-Stage-Eus2-Plb-Rule          deleted
     -   ├─ azure:lb:Probe                      Api-B-Http-4422-Up-Stage-Eus2-P-Probe     deleted
     -   ├─ azure:lb:BackendAddressPool         Api-B-Stage-Eus2-Plb-Pool                 deleted
     -   ├─ azure:network:NetworkSecurityRule   Api-B-Tcp-443-Allow-Stage-Eus2-Nsg-Rule   deleted
     -   ├─ azure:network:NetworkSecurityRule   Api-B-Tcp-80-Allow-Stage-Eus2-Nsg-Rule    deleted
     -   ├─ azure:lb:LoadBalancer               Api-B-Stage-Eus2-Plb                      deleted
     -   ├─ azure:network:NetworkSecurityRule   Api-B-Any-3444-Allow-Stage-Eus2-Nsg-Rule  deleted
     -   ├─ azure:network:NetworkSecurityGroup  Api-B-Stage-Eus2-Nsg                      deleted
     -   ├─ azure:network:PublicIp              Api-B-Stage-Eus2-Pip                      deleted
     -   └─ azure:core:ResourceGroup            Api-B-Stage-Eus2-Rg                       deleted
    
    Diagnostics:
      pulumi:pulumi:Stack (compute-api-compute-api-b-stage-eus2):
        (node:99) UnhandledPromiseRejectionWarning: Error: Failed to load gRPC binary module because it was not installed for the current system
        Expected directory: node-v64-linux-x64-glibc
        Found: [node-v64-darwin-x64-unknown]
        This problem can often be fixed by running "npm rebuild" on the current system
        Original error: Cannot find module '/data/pulumi/node_modules/grpc/src/node/extension_binary/node-v64-linux-x64-glibc/grpc_node.node'
            at Object.<anonymous> (/data/pulumi/node_modules/grpc/src/grpc_extension.js:53:17)
            at Module._compile (internal/modules/cjs/loader.js:689:30)
            at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
            at Module.load (internal/modules/cjs/loader.js:599:32)
            at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
            at Function.Module._load (internal/modules/cjs/loader.js:530:3)
            at Module.require (internal/modules/cjs/loader.js:637:17)
            at require (internal/modules/cjs/helpers.js:22:18)
            at Object.<anonymous> (/data/pulumi/node_modules/grpc/src/client_interceptors.js:144:12)
            at Module._compile (internal/modules/cjs/loader.js:689:30)
        (node:99) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
        (node:99) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
    
    Resources:
        - 12 deleted
    
    Duration: 3m6s
````

--- a/sdk/nodejs/cmd/run/index.ts
+++ b/sdk/nodejs/cmd/run/index.ts
@@ -23,7 +23,7 @@ const uncaughtHandler = (err: Error) => {
     if (!programRunning) {
         console.error(err.stack || err.message);
     }
-}
+};
 process.on("uncaughtException", uncaughtHandler);
 process.on("unhandledRejection", uncaughtHandler);
 process.on("exit", (code: number) => {

--- a/sdk/nodejs/cmd/run/index.ts
+++ b/sdk/nodejs/cmd/run/index.ts
@@ -12,7 +12,28 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// As the very first thing we do, ensure that we're connected to v8's inspector API.  We need to do
+// The very first thing we do is set up unhandled exception and rejection hooks to ensure that these events cause us to
+// exit with a non-zero code. It is criticially important that we do this early: if we do not, unhandled rejections in
+// particular may cause us to exit with a 0 exit code, which will trick the engine into thinking that the program ran
+// successfully. This can cause the engine to decide to delete all of a stack's resources.
+let uncaught: Error | undefined;
+let programRunning = false;
+const uncaughtHandler = (err: Error) => {
+    uncaught = err;
+    if (!programRunning) {
+        console.error(err.stack || err.message);
+    }
+}
+process.on("uncaughtException", uncaughtHandler);
+process.on("unhandledRejection", uncaughtHandler);
+process.on("exit", (code: number) => {
+    // If we don't already have an exit code, and we had an unhandled error, exit with a non-success.
+    if (code === 0 && uncaught) {
+        process.exitCode = 1;
+    }
+});
+
+// As the second thing we do, ensure that we're connected to v8's inspector API.  We need to do
 // this as some information is only sent out as events, without any way to query for it after the
 // fact.  For example, we want to keep track of ScriptId->FileNames so that we can appropriately
 // report errors for Functions we cannot serialize.  This can only be done (up to Node11 at least)
@@ -88,7 +109,10 @@ function main(args: string[]): void {
 
     // Ensure that our v8 hooks have been initialized.  Then actually load and run the user program.
     v8Hooks.isInitializedAsync().then(() => {
-        require("./run").run(argv);
+        require("./run").run(argv, () => {
+            programRunning = true;
+        });
+        programRunning = false;
     });
 }
 

--- a/sdk/nodejs/cmd/run/run.ts
+++ b/sdk/nodejs/cmd/run/run.ts
@@ -117,7 +117,7 @@ function reportModuleLoadFailure(program: string, error: Error): never {
     return process.exit(1);
 }
 
-export function run(argv: minimist.ParsedArgs): void {
+export function run(argv: minimist.ParsedArgs, programStarted: () => void): void {
     // If there is a --pwd directive, switch directories.
     const pwd: string | undefined = argv["pwd"];
     if (pwd) {
@@ -158,7 +158,6 @@ export function run(argv: minimist.ParsedArgs): void {
     process.argv = [ process.argv[0], process.argv[1], ...programArgs ];
 
     // Set up the process uncaught exception, unhandled rejection, and program exit handlers.
-    let uncaught: Error | undefined;
     const errorSet = new Set<Error>();
 
     const uncaughtHandler = (err: Error) => {
@@ -190,22 +189,13 @@ export function run(argv: minimist.ParsedArgs): void {
             log.error(`Running program '${program}' failed with an unhandled exception:`);
             log.error(defaultMessage);
         }
-
-        // Remember that we failed with an error.  Don't quit just yet so we have a chance to drain the message loop.
-        uncaught = err;
     };
 
     process.on("uncaughtException", uncaughtHandler);
     process.on("unhandledRejection", uncaughtHandler);
+    process.on("exit", runtime.disconnectSync);
 
-    process.on("exit", (code: number) => {
-        runtime.disconnectSync();
-
-        // If we don't already have an exit code, and we had an unhandled error, exit with a non-success.
-        if (code === 0 && uncaught) {
-            process.exitCode = 1;
-        }
-    });
+    programStarted();
 
     // Construct a `Stack` resource to represent the outputs of the program.
     runtime.runInPulumiStack(() => {

